server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -121,6 +121,7 @@
     private final List<IndexEventListener> buildInIndexListener;
     private final PrimaryReplicaSyncer primaryReplicaSyncer;
     private final Consumer<ShardId> globalCheckpointSyncer;
+    private volatile ClusterState clusterState; // the latest applied cluster state
```

Comment:
I'm not a fan of this. How about doing the following instead: We first fetch the cluster state with no metadata version requested. In case where the mapping version is less than the mapping version we want, we issue another cluster state request, this time asking for the cluster state with the next metadata version (i.e. current metadata version + 1)? This means that we potentially still ask twice for the cluster state, but I find that still preferable to all these changes here.